### PR TITLE
shared memory allocator documenation

### PR DESCRIPTION
**Pull request by psychocoderHPC**
add more documentation to avaoid wrong usage of the macro `PMACC_SMEM()`

### CODE REVIEW

File: src/libPMacc/include/memory/shared/Allocate.hpp [hunk 1]
```diff
@@ -105,6 +105,27 @@ namespace shared
 
 /** allocate shared memory
  *
+ * Do not use this macro within a factory method. This macro contains a pre compiler
```
Comment (ax3l) [inline src/libPMacc/include/memory/shared/Allocate.hpp]: worth a `@warning` I think :)

File: src/libPMacc/include/memory/shared/Allocate.hpp [hunk 2]
```diff
@@ -105,6 +105,27 @@ namespace shared
...
+ * counter which is only increased once per source code line.
+ * Please do also not assume that this macro can be used to create aliased pointer
```
Comment (ax3l) [inline src/libPMacc/include/memory/shared/Allocate.hpp]: You can even write: `You must not use this macro to create an aliased pointer to shared memory.`

File: src/libPMacc/include/memory/shared/Allocate.hpp [hunk 3]
```diff
@@ -105,6 +105,27 @@ namespace shared
...
+ * to shared memory!
+ *
+ * @code
+ * // THIS IS A EXAMPLE HOW `PMACC_SMEM` SHOULD NOT BE USED
```
Comment (ax3l) [inline src/libPMacc/include/memory/shared/Allocate.hpp]: an example

really a wrong-usage but not a a right-usage example?

File: src/libPMacc/include/memory/shared/Allocate.hpp [hunk 4]
```diff
@@ -105,6 +105,27 @@ namespace shared
...
+ *     {
+ *         // this macro points always to the same memory address
+ *         // even if this method is called twice
+ *         PMACC_SMEM( sharedMem, int );
+ *         return sharedMem;
+ *     }
+ * };
+ * @endcode
+ *
+ * Use `PMacc::memory::shared::allocate<id,type>()` instead.
```
Comment (ax3l) [inline src/libPMacc/include/memory/shared/Allocate.hpp]: how? code example would be great :)

**Comment by psychocoderHPC** [comment]
updated

File: src/libPMacc/include/memory/shared/Allocate.hpp [hunk 5]
```diff
@@ -105,6 +105,38 @@ namespace shared
...
+ *     }
+ * };
+ *
+ * // RIGHT USAGE
+ * template< uint32_t T_id >
+ * struct Factory
+ * {
+ *     int& getSharedMem( )
+ *     {
+ *         // create new shared memory for each `T_id`
```
Comment (PrometheusPi) [inline src/libPMacc/include/memory/shared/Allocate.hpp]: QUESTION:
What do you mean with *for each `T_id`* - there is just **one** `T_id`?

Comment (ax3l) [inline src/libPMacc/include/memory/shared/Allocate.hpp]: for each specialization of the class with always an other T_id. I find the wording ok

Comment (PrometheusPi) [inline src/libPMacc/include/memory/shared/Allocate.hpp]: Okay - that makes sense

**Pull request merged**